<a href="https://colab.research.google.com/github/valeriafigueredo-data/Mineracao_de_Texto_e_na_Web/blob/main/4_Desafio_TFIDF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**DESAFIO 2**

**DISCENTE: Valéria Cristina A. R. de Figueredo**

### Instruções

* Rode o código da célula abaixo para obter os textos que serão usados no desafio, utilize entre 10 a 20 textos da base mudando o N.
* Crie um DataFrame com o TF, IDF e TF-IDF de todos os termos
* O que os 5 termos com o maior TF-IDF significam na tua opinião? Porque eles são os maiores?
**Não esqueça dos passos antes de ter um termo/token**

In [ ]:
from sklearn.datasets import fetch_20newsgroups

# Descomente a linha de baixo se quiser ver mais detalhes do dataset
# print(fetch_20newsgroups.__doc__)
N = 20
raw_documents= fetch_20newsgroups(subset='train', return_X_y=True)[0]
for idx, doc in enumerate(raw_documents[:N]):
  print(f"========== Doc {idx} ==========")
  print(doc)

========== Doc 0 ==========
From: lerxst@wam.umd.edu (where's my thing)
Subject: WHAT car is this!?
Nntp-Posting-Host: rac3.wam.umd.edu
Organization: University of Maryland, College Park
Lines: 15

 I was wondering if anyone out there could enlighten me on this car I saw
the other day. It was a 2-door sports car, looked to be from the late 60s/
early 70s. It was called a Bricklin. The doors were really small. In addition,
the front bumper was separate from the rest of the body. This is 
all I know. If anyone can tellme a model name, engine specs, years
of production, where this car is made, history, or whatever info you
have on this funky looking car, please e-mail.

Thanks,
- IL
   ---- brought to you by your neighborhood Lerxst ----





========== Doc 1 ==========
From: guykuo@carson.u.washington.edu (Guy Kuo)
Subject: SI Clock Poll - Final Call
Summary: Final call for SI clock reports
Keywords: SI,acceleration,clock,upgrade
Article-I.D.: shelley.1qvfo9INNc3s
Organization: Universit

In [ ]:
pip install nltk

# **1. Bibliotecas**

In [ ]:
# Bibliotecas
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import re
import nltk
from nltk.corpus import stopwords

# **2. Pré-Processamento**

In [ ]:

# Stop words do NLTK
nltk.download('stopwords')

# Lista de stop words em inglês
stop_words = set(stopwords.words('english'))

def preprocess(text):

    # Substitui quebra de linha por espaço
    text = re.sub(r"\n", " ",text)
    # Remove caracteres especiais
    text = re.sub(r'[^\w\s]', '', text)
    # Converte para minúsculas
    text = text.lower()
    # Remove espaços em branco extras
    text = re.sub(r'\s+', ' ', text)
    # Remove stop words
    text = ' '.join([word for word in text.split() if word not in stop_words])
    text = text.strip()
    return text.strip()

# Pré-processamento dos documentos
processed_documents = [preprocess(doc) for doc in raw_documents[:N]]

# Exibição dos documentos pré-processados
for idx, doc in enumerate(processed_documents):
    print(f"========== Doc {idx} ==========")
    print(doc)

========== Doc 0 ==========
lerxstwamumdedu wheres thing subject car nntppostinghost rac3wamumdedu organization university maryland college park lines 15 wondering anyone could enlighten car saw day 2door sports car looked late 60s early 70s called bricklin doors really small addition front bumper separate rest body know anyone tellme model name engine specs years production car made history whatever info funky looking car please email thanks il brought neighborhood lerxst
========== Doc 1 ==========
guykuocarsonuwashingtonedu guy kuo subject si clock poll final call summary final call si clock reports keywords siaccelerationclockupgrade articleid shelley1qvfo9innc3s organization university washington lines 11 nntppostinghost carsonuwashingtonedu fair number brave souls upgraded si clock oscillator shared experiences poll please send brief message detailing experiences procedure top speed attained cpu rated speed add cards adapters heat sinks hour usage per day floppy disk functionalit

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
# Tokenização dos documentos pré-processados
tokenized_documents = [word_tokenize(doc) for doc in processed_documents]

# Exibição dos documentos tokenizados
for idx, tokens in enumerate(tokenized_documents):
    print(f"========== Doc {idx} ==========")
    print(tokens)

========== Doc 0 ==========
['lerxstwamumdedu', 'wheres', 'thing', 'subject', 'car', 'nntppostinghost', 'rac3wamumdedu', 'organization', 'university', 'maryland', 'college', 'park', 'lines', '15', 'wondering', 'anyone', 'could', 'enlighten', 'car', 'saw', 'day', '2door', 'sports', 'car', 'looked', 'late', '60s', 'early', '70s', 'called', 'bricklin', 'doors', 'really', 'small', 'addition', 'front', 'bumper', 'separate', 'rest', 'body', 'know', 'anyone', 'tellme', 'model', 'name', 'engine', 'specs', 'years', 'production', 'car', 'made', 'history', 'whatever', 'info', 'funky', 'looking', 'car', 'please', 'email', 'thanks', 'il', 'brought', 'neighborhood', 'lerxst']
========== Doc 1 ==========
['guykuocarsonuwashingtonedu', 'guy', 'kuo', 'subject', 'si', 'clock', 'poll', 'final', 'call', 'summary', 'final', 'call', 'si', 'clock', 'reports', 'keywords', 'siaccelerationclockupgrade', 'articleid', 'shelley1qvfo9innc3s', 'organization', 'university', 'washington', 'lines', '11', 'nntppostingho

In [ ]:
# Remoção de tokens não alfabéticos
filtered_tokens = [[word for word in tokens if word.isalpha()] for tokens in tokenized_documents]

# Exibição dos tokens filtrados
for idx, tokens in enumerate(filtered_tokens):
    print(f"========== Filtered Tokens for Doc {idx} ==========")
    print(tokens)

========== Filtered Tokens for Doc 0 ==========
['lerxstwamumdedu', 'wheres', 'thing', 'subject', 'car', 'nntppostinghost', 'organization', 'university', 'maryland', 'college', 'park', 'lines', 'wondering', 'anyone', 'could', 'enlighten', 'car', 'saw', 'day', 'sports', 'car', 'looked', 'late', 'early', 'called', 'bricklin', 'doors', 'really', 'small', 'addition', 'front', 'bumper', 'separate', 'rest', 'body', 'know', 'anyone', 'tellme', 'model', 'name', 'engine', 'specs', 'years', 'production', 'car', 'made', 'history', 'whatever', 'info', 'funky', 'looking', 'car', 'please', 'email', 'thanks', 'il', 'brought', 'neighborhood', 'lerxst']
========== Filtered Tokens for Doc 1 ==========
['guykuocarsonuwashingtonedu', 'guy', 'kuo', 'subject', 'si', 'clock', 'poll', 'final', 'call', 'summary', 'final', 'call', 'si', 'clock', 'reports', 'keywords', 'siaccelerationclockupgrade', 'articleid', 'organization', 'university', 'washington', 'lines', 'nntppostinghost', 'carsonuwashingtonedu', 'fair'

# **3. TF**

In [ ]:
import pandas as pd
import numpy as np

# Criação de um conjunto de todas as palavras únicas
unique_tokens = set(token for tokens in filtered_tokens for token in tokens)

# Criação de um DataFrame para armazenar contagens e TF
df = pd.DataFrame(data=np.zeros((2, len(unique_tokens))), columns=list(unique_tokens), index=["N", "TF"])

# Conta a frequência de cada token
for tokens in filtered_tokens:
    for word in tokens:
        if word in df.columns:
            df.at["N", word] += 1

# Calcula TF (Term Frequency)
total_tokens = df.loc["N"].sum()  # Total de todas as contagens
for token in df.columns:
    df.at["TF", token] = df.at["N", token] / total_tokens if total_tokens > 0 else 0

# Exibi o DataFrame com as contagens e TF
print(df)

    motors  deleted  friends   cubic     ftp  warmly    bike   tough  \
N   2.0000   2.0000   1.0000  1.0000  1.0000  1.0000  2.0000  1.0000   
TF  0.0006   0.0006   0.0003  0.0003  0.0003  0.0003  0.0006  0.0003   

    destruction   keith  ...     tsi  insert    sale  boston     srt   three  \
N      7.000000  1.0000  ...  1.0000  1.0000  2.0000  1.0000  1.0000  2.0000   
TF     0.002101  0.0003  ...  0.0003  0.0003  0.0006  0.0003  0.0003  0.0006   

     sound    fort  trooper   cards  
N   1.0000  2.0000   1.0000  1.0000  
TF  0.0003  0.0006   0.0003  0.0003  

[2 rows x 1737 columns]


In [ ]:
# Transpor o DataFrame
transposed_df = df.T  # As linhas se tornam colunas e vice-versa
transposed_df

N      TF
motors   2.0  0.0006
deleted  2.0  0.0006
friends  1.0  0.0003
cubic    1.0  0.0003
ftp      1.0  0.0003
...      ...     ...
three    2.0  0.0006
sound    1.0  0.0003
fort     2.0  0.0006
trooper  1.0  0.0003
cards    1.0  0.0003

[1737 rows x 2 columns]

In [ ]:
# Ordenamento do DataFrame transposto por TF em ordem decrescente
sorted_df = transposed_df.sort_values(by='TF', ascending=False)
sorted_df

N        TF
would         22.0  0.006605
subject       20.0  0.006004
lines         20.0  0.006004
organization  20.0  0.006004
im            19.0  0.005704
...            ...       ...
called         1.0  0.000300
ancestors      1.0  0.000300
brother        1.0  0.000300
ryan           1.0  0.000300
cards          1.0  0.000300

[1737 rows x 2 columns]

In [ ]:
# 5 Tokens com maior TF
# Encontrar os 5 termos com o maior TF
top_terms = df.loc["TF"].nlargest(5)

# Exibi os 5 termos com maior TF
print("Top 5 termos com maior TF:")
print(top_terms)

Top 5 termos com maior TF:
would           0.006605
organization    0.006004
subject         0.006004
lines           0.006004
im              0.005704
Name: TF, dtype: float64


# **4. IDF**

In [ ]:
import math

# Calcular o número total de documentos
total_documents = len(filtered_tokens)

# Criar um dicionário para armazenar o IDF
idf_values = {}

# Calcular IDF para cada token
for token in df.columns:
    # n_t é o número de documentos que contêm o token
    n_t = df.at["N", token]  # Aqui está o número total de ocorrências do token

    # Calcular IDF
    idf_values[token] = math.log10(total_documents / n_t) if n_t > 0 else 0

# Adicionar os valores de IDF ao DataFrame
idf_df = pd.DataFrame.from_dict(idf_values, orient='index', columns=['IDF'])
idf_df.index.name = 'Word'

print("\nDataFrame de IDF:")
print(idf_df)


DataFrame de IDF:
             IDF
Word            
motors   1.00000
deleted  1.00000
friends  1.30103
cubic    1.30103
ftp      1.30103
...          ...
three    1.00000
sound    1.30103
fort     1.00000
trooper  1.30103
cards    1.30103

[1737 rows x 1 columns]


In [ ]:
# Ordenamento do DataFrame transposto por IDF em ordem decrescente
sorted_df2 = idf_df.sort_values(by='IDF', ascending=False)
sorted_df2

IDF
Word                  
dan           1.301030
start         1.301030
finals        1.301030
heres         1.301030
note          1.301030
...                ...
im            0.022276
lines         0.000000
organization  0.000000
subject       0.000000
would        -0.041393

[1737 rows x 1 columns]

In [ ]:
# 5 Tokens com maior IDF
top_5_idf = idf_df.nlargest(5, 'IDF')

# Exibir os 5 termos com maior IDF
print("\nTop 5 termos com maior IDF:")
print(top_5_idf)




Top 5 termos com maior IDF:
             IDF
Word            
friends  1.30103
cubic    1.30103
ftp      1.30103
warmly   1.30103
tough    1.30103


# **5. TF-IDF**

In [ ]:

# Criar um DataFrame para armazenar o TF-IDF
tf_idf_df = pd.DataFrame(index=df.columns, columns=['TF-IDF'])

# Calcular o TF-IDF
for token in df.columns:
    tf_idf_df.at[token, 'TF-IDF'] = df.at["TF", token] * idf_df.at[token, 'IDF']

# Exibir o DataFrame com os resultados de TF-IDF
print(tf_idf_df)



           TF-IDF
motors     0.0006
deleted    0.0006
friends  0.000391
cubic    0.000391
ftp      0.000391
...           ...
three      0.0006
sound    0.000391
fort       0.0006
trooper  0.000391
cards    0.000391

[1737 rows x 1 columns]


# **6. Dataframe com todos Indicadores**

In [ ]:
# Exibição do DataFrame com os resultados de TF, IDF e TF-IDF

# Combinar os DataFrames em um único DataFrame
combined_df = pd.concat([df.loc["TF"], idf_df['IDF'], tf_idf_df['TF-IDF']], axis=1)
combined_df.columns = ['TF', 'IDF', 'TF-IDF']

# Ordenar o DataFrame pelo valor de TF-IDF do maior para o menor
combined_df = combined_df.sort_values(by='TF-IDF', ascending=False)

# Exibir o DataFrame combinado e ordenado
print("DataFrame Combinado de TF, IDF e TF-IDF (ordenado pelo TF-IDF):")
print(combined_df)

DataFrame Combinado de TF, IDF e TF-IDF (ordenado pelo TF-IDF):
                    TF       IDF    TF-IDF
could         0.002101  0.455932  0.000958
redesign      0.002101  0.455932  0.000958
use           0.002101  0.455932  0.000958
thats         0.002101  0.455932  0.000958
board         0.002101  0.455932  0.000958
...                ...       ...       ...
im            0.005704  0.022276  0.000127
subject       0.006004  0.000000       0.0
organization  0.006004  0.000000       0.0
lines         0.006004  0.000000       0.0
would         0.006605 -0.041393 -0.000273

[1737 rows x 3 columns]


In [ ]:
# Converter a coluna 'TF-IDF' para float
tf_idf_df['TF-IDF'] = tf_idf_df['TF-IDF'].astype(float)

# 5 Tokens com maior TF-IDF
top_5_tfidf = tf_idf_df.nlargest(5, 'TF-IDF')

# Exibir os 5 termos com maior TF-IDF
print("Top 5 termos com maior TF-IDF:")
print(top_5_tfidf)

Top 5 termos com maior TF-IDF:
               TF-IDF
destruction  0.000958
performance  0.000958
use          0.000958
station      0.000958
board        0.000958


# **7. Conclusão**



No presente trabalho foi elaborado um DataFrame com o TF, IDF e TF-IDF de todos os termos (tokens). Para isso foi realizada a limpeza dos dados, o seu pré-processamento para eliminar sujeiras, caracteres, espaços extras, enfim, dados não essenciais para a análise.

O TF (Frequência do Termo) significa a frequência de um termo em um documento específico. Quanto mais vezes um termo aparece em um documento, mais importante ele é considerado nesse documento.

O IDF (Frequência Inversa de Documento) quantifica a importância de um termo em relação a um conjunto de documentos (ou corpus).Isto porque termos que aparecem frequentemente em muitos documentos não são tão informativos como "e", "ou", "mas", de modo que não agregam valor significativo à análise.

O TF-IDF, por sua vez, significa Frequência do Termo - Frequência Inversa do Documento. É uma medida usada para avaliar a importância de uma palavra em um documento dentro de um conjunto de documentos (ou corpus). Ele combina as duas informações: TF e IDF. Logo, são termos que ocorrem frequentemente em um documento mas de forma rara nos demais.

Sobre os valores da medida TF-IDF, não tem um limite fixo, variam entre valores negativos, 0 a valores superiores a 1, dependendo da frequência do termo e de sua distribuição nos documentos. TF- IDF = 0 significa que o termo não aparece no documento ou que é uma palavra muito comum que aparece em todos os documentos, resultando em um IDF baixo. E quanto mais afastado de 0, positivo, indica que o termo é relevante para o documento e ao corpus como um todo.

O Top 5 termos com maior TF foi would, organization, subject, lines e im. Isso porque tais termos têm alta frequência nos documentos. Por exemplo, subject, lines  e organization estão na parte do cabeçalho dos e-mails e todos textos são emails.

O Top 5 termos com maior TF-IDF foram destruction, performance, use, station e board  (0.000958). Isso significa que apareceram bastante em um documento (texto) e pouco nos demais. POr exemplo, o termo destruction apareceu bastante no texto 6 (DOC 5), abordando sobre armas de destruição em massa e o seu impacto na sociedade. Entretanto, para os demais textos (DOC) tal termo apareceu de forma irrelevante.
